File: coloredwave.cpp
#include "coloredwave.h"

ColoredWave::ColoredWave() {}

Color ColoredWave::getColor(int index) const {
    Color ret;
    if(index < colori.size()) {
        ret=*colori[index];
    } return ret;
}


void ColoredWave::setColor(Color color, int index, bool insert) {
    Color* aux = new Color(color);
    if(index < 0) {
        colori.push_back(aux);
    } else if (insert < waveLenght()){
        if(insert){
            colori.insert(index, aux);
        } else {
            colori.replace(index, aux);
        }
    }
}

void ColoredWave::pushPoint(double val, Color color, int index, bool insert) {
    this->setVAmp(val, index, insert);
    this->setColor(color, index, insert);
}

void ColoredWave::removePoint(const int index) {
    if (index < colori.size()) {
        colori.remove(index);
        v.remove(index);
    }
}

int ColoredWave::waveLenght() const{
    return colori.size();
}

double ColoredWave::getAmpVal(const int index) const {
    return getVAmp(index);
}

Color ColoredWave::getAmpColor(const int index) const {
    return getColor(index);
}

QString ColoredWave::getString(){
    QString ret("("), comma;
    QMutableVectorIterator<Color*> i(colori);
    while (i.hasNext()) {
        ret.append(comma);
        ret.append(Color::ConvertRGBtoHex(*i.next());
        comma = ",";
    }
    ret.append(")");
    return ret;
}

ColoredWave ColoredWave::operator+(const ColoredWave& x){
    ColoredWave* aux = new ColoredWave();
    int maxLenght = qMax(this->waveLenght(), x.waveLenght());

    Color res;

    for(int i=0; i<maxLenght; ++i) {
        res = this->getColor(i) + x.getColor(i);
        aux->pushPoint(this->getAmpVal(i) + x.getAmpVal(i), res);
    }
    return *aux;
}
/*ColoredWave ColoredWave::operator-(const ColoredWave& x){
    ColoredWave* aux = new ColoredWave();
    int maxLenght = qMax(this->waveLenght(), x.waveLenght());

    Color res;

    for(int i=0; i<maxLenght; ++i) {
        res = this->getColor(i) - x.getColor(i);
        aux->pushPoint(this->getAmpVal(i) - x.getAmpVal(i), res);
    }
    return *aux;
}
ColoredWave ColoredWave::operator*(const ColoredWave& x){
    ColoredWave* aux = new ColoredWave();
    int maxLenght = qMax(this->waveLenght(), x.waveLenght());

    Color res;

    for(int i=0; i<maxLenght; ++i) {
        res = this->getColor(i) * x.getColor(i);
        aux->pushPoint(this->getAmpVal(i) * x.getAmpVal(i), res);
    }
    return *aux;
}
ColoredWave ColoredWave::operator/(const ColoredWave& x){
    ColoredWave* aux = new ColoredWave();
    int maxLenght = qMax(this->waveLenght(), x.waveLenght());

    Color res;

    for(int i=0; i<maxLenght; ++i) {
        res = this->getColor(i) / x.getColor(i);
        aux->pushPoint(this->getAmpVal(i) / x.getAmpVal(i), res);
    }
    return *aux;
}   */

File: java/src/kalk/wave/ColoredWave.java
package kalk.wave;

import java.util.ArrayList;

import static java.lang.Integer.max;

public class ColoredWave extends Wave {

    private ArrayList<Color> colori = new ArrayList<Color>();

    public ColoredWave() {}

    public Color getColore(Integer index) {
        Color ret  = new Color();
        if (index < colori.size()) {
            ret = colori.get(index);
        } return ret;
    }

    public void setColor(Color color, Integer index, boolean insert){
        Color aux = color;
        if (index < 0) {
            colori.add(aux);
        }else if (index < waveLenght()) {
            if (insert) {
                colori.add(index, aux);
            } else {
                colori.set(index, aux);
            }
        }
    }

    public void pushPoint(Double x, Color color, Integer index, boolean insert){
        setVAmp(x, index, insert);
        setColor(color, index, insert);
    }

    public void pushPoint (Double x, Color color, Integer index){
        pushPoint(x, color, index, false);
    }

    public void pushPoint (Double x, Color color){
        pushPoint(x, color, -1, false);
    }

    public void removePoint(Integer  index) {
        if (index < colori.size()) {
            colori.remove(index.intValue());
            v.remove(index.intValue());
        }
    }

    public Integer waveLenght() {
        return colori.size();
    }

    public Double getAmpVal(Integer index) {
        return getVAmp(index);
    }

    public Color getAmpColor(Integer index) {
        return getColore(index);
    }

    public ColoredWave somma(ColoredWave x) {
        ColoredWave aux = new ColoredWave();
        Integer maxLenght = max(waveLenght(), x.waveLenght());
        Color res = new Color();
        for(int i=0; i<maxLenght; ++i) {
            res.somma(getColore(i), x.getColore(i));
            aux.pushPoint(getAmpVal(i) + x.getAmpVal(i), res);
        }
        return aux;
    }

    public ColoredWave sottrazione(ColoredWave x) {
        ColoredWave aux = new ColoredWave();
        Integer maxLenght = max(waveLenght(), x.waveLenght());
        Color res = new Color();
        for(int i=0; i<maxLenght; ++i) {
            res.sottrazione(getColore(i), x.getColore(i));
            aux.pushPoint(getAmpVal(i) - x.getAmpVal(i), res);
        }
        return aux;
    }

    public ColoredWave moltiplicazione(ColoredWave x) {
        ColoredWave aux = new ColoredWave();
        Integer maxLenght = max(waveLenght(), x.waveLenght());
        Color res = new Color();;
        for(int i=0; i<maxLenght; ++i) {
            res.moltiplicazione(getColore(i), x.getColore(i));
            aux.pushPoint(getAmpVal(i) * x.getAmpVal(i), res);
        }
        return aux;
    }

    public ColoredWave divisione(ColoredWave x) {
        ColoredWave aux = new ColoredWave();
        Integer maxLenght = max(waveLenght(), x.waveLenght());
        Color res = new Color();
        for(int i=0; i<maxLenght; ++i) {
            res.divisione(getColore(i), x.getColore(i));
            aux.pushPoint(getAmpVal(i) / x.getAmpVal(i), res);
        }
        return aux;
    }
}
